airflow/providers/amazon/aws/operators/redshift_cluster.py

```diff
@@ -524,64 +527,52 @@ def __init__(
         aws_conn_id: str = "aws_default",
         deferrable: bool = False,
         poll_interval: int = 10,
+        max_attempts: int = 15,
         **kwargs,
     ):
         super().__init__(**kwargs)
         self.cluster_identifier = cluster_identifier
         self.aws_conn_id = aws_conn_id
         self.deferrable = deferrable
+        self.max_attempts = max_attempts
         self.poll_interval = poll_interval
-        # These parameters are added to address an issue with the boto3 API where the API
+        # These parameters are used to address an issue with the boto3 API where the API
         # prematurely reports the cluster as available to receive requests. This causes the cluster
         # to reject initial attempts to pause the cluster despite reporting the correct state.
         self._attempts = 10
         self._attempt_interval = 15
 
     def execute(self, context: Context):
         redshift_hook = RedshiftHook(aws_conn_id=self.aws_conn_id)
+        while self._attempts >= 1:
+            try:
+                redshift_hook.get_conn().pause_cluster(ClusterIdentifier=self.cluster_identifier)
+                break
+            except redshift_hook.get_conn().exceptions.InvalidClusterStateFault as error:
+                self._attempts = self._attempts - 1
 
+                if self._attempts > 0:
+                    self.log.error("Unable to pause cluster. %d attempts remaining.", self._attempts)
+                    time.sleep(self._attempt_interval)
+                else:
+                    raise error
         if self.deferrable:
             self.defer(
-                timeout=self.execution_timeout,
-                trigger=RedshiftClusterTrigger(
-                    task_id=self.task_id,
+                trigger=RedshiftPauseClusterTrigger(
+                    cluster_identifier=self.cluster_identifier,
                     poll_interval=self.poll_interval,
+                    max_attempts=self.max_attempts,
```

Comment:
so, with triggers, there's already a timeout.  why don't we continue to use that? though it shouldn't be execution_timeout -- that's the timeout for the task itself. but you could calculate it from max_attempts*poll interval.  and then your trigger doesn't need a max_attempts param it can just loop until it times out.

one problem with using max_attempts in trigger is if trigger is killed and is respawned it will reset -- not so if we use timeout.

Reply:
Using the max attempts vs the timeout was essentially a design choice due to how the boto waiter implements things. The boto waiter defines `MaxAttempts` and `Delay` so we decided to go with those parameters. Initially, when we had our own manual waiter, we were going with a `waiter_countdown` [parameter](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/operators/emr.py#L176) which would map to the `timeout` on the Trigger. But we are in the process of deprecating that parameter. ( As an aside, we calculate `waiter_max_attempts` as `waiter_max_attempts = waiter_countdown // waiter_check_interval_seconds`).

>one problem with using max_attempts in trigger is if trigger is killed and is respawned it will reset -- not so if we use timeout.

This is a good point. We could solve this by making the `attempt` number as part of the kwargs of the trigger - which would save the value if it was killed at any point. What do you think?

Reply:
I don't see how that wolud address the issue.  If it's part of the kwargs, and trigger killed, when new trigger spawned, same kwargs, same behavior.  Best to use the timeout feature for triggers.

Reply:
alright, that sounds like a fair plan. I'll make the changes

Reply:
Thinking about it a bit more, does Airflow support emitting a TriggerEvent on Trigger timeout? There is a `cleanup` method that executes after `run` but can we emit a TriggerEvent from there? Or is there a better way?

Reply:
I ended up going with adding a timeout, as you suggested. But I also left the code to track the number of attempts in as well. The reason is, when a Trigger times out, it does not yield a `TriggerEvent`. This makes the logs slightly confusing, because it seems like the Trigger succeeded. 
```
[2023-05-08, 02:19:30 UTC] {credentials.py:617} INFO - Found credentials in shared credentials file: ~/.aws/credentials
[2023-05-08, 02:19:38 UTC] {redshift_cluster.py:203} INFO - Status of cluster is pausing
[2023-05-08, 02:19:57 UTC] {redshift_cluster.py:203} INFO - Status of cluster is pausing
[2023-05-08, 02:20:15 UTC] {redshift_cluster.py:203} INFO - Status of cluster is pausing
[2023-05-08, 02:20:28 UTC] {taskinstance.py:1144} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: example_redshift.pause_cluster manual__2023-05-08T02:11:38.809301+00:00 [queued]>
[2023-05-08, 02:20:28 UTC] {taskinstance.py:1144} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: example_redshift.pause_cluster manual__2023-05-08T02:11:38.809301+00:00 [queued]>
[2023-05-08, 02:20:28 UTC] {taskinstance.py:1343} INFO - Resuming after deferral
[2023-05-08, 02:20:28 UTC] {taskinstance.py:1364} INFO - Executing <Task(RedshiftPauseClusterOperator): pause_cluster> on 2023-05-08 02:11:38.809301+00:00
[2023-05-08, 02:20:28 UTC] {standard_task_runner.py:57} INFO - Started process 328 to run task
[2023-05-08, 02:20:28 UTC] {standard_task_runner.py:84} INFO - Running: ['***', 'tasks', 'run', 'example_redshift', 'pause_cluster', 'manual__2023-05-08T02:11:38.809301+00:00', '--job-id', '17', '--raw', '--subdir', 'DAGS_FOLDER/example_redshift.py', '--cfg-path', '/tmp/tmp582aoom9']
[2023-05-08, 02:20:28 UTC] {standard_task_runner.py:85} INFO - Job 17: Subtask pause_cluster
[2023-05-08, 02:20:28 UTC] {task_command.py:410} INFO - Running <TaskInstance: example_redshift.pause_cluster manual__2023-05-08T02:11:38.809301+00:00 [running]> on host 9ed7c7091244
[2023-05-08, 02:20:28 UTC] {taskinstance.py:1863} ERROR - Task failed with exception
airflow.exceptions.TaskDeferralError: Trigger/execution timeout
[2023-05-08, 02:20:28 UTC] {taskinstance.py:1382} INFO - Marking task as FAILED. dag_id=example_redshift, task_id=pause_cluster, execution_date=20230508T021138, start_date=20230508T021920, end_date=20230508T022028
```
I felt this was not a good user experience, so I decided to add the timeout along with the number of attempts. This way, under normal circumstances, a Trigger would fail after attempting the max number of attempts, and the Trigger would yield a `TriggerEvent` with an error message explaining the error. In the case the Trigger is killed and respawned, the timeout from the operator will persist and fail the Trigger at the correct time.

Reply:
I see what you mean... 

Your concern is that after `[2023-05-08, 02:20:15 UTC]` there's no message saying "trigger timeout"

Perhaps that's something we should fix.